### PR TITLE
Lutece 2270: Asynchronous Generation of Files

### PR DESCRIPTION
**Pull request by glepine1**

### CODE REVIEW

File: src/java/fr/paris/lutece/portal/service/tempfile/TemporaryFileGeneratorService.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+        private PhysicalFile createPhysicalFile( Path generatedFile ) throws IOException
+        {
+            PhysicalFile physicalFile = new PhysicalFile( );
+            if ( _generator.isZippable( ) )
+            {
+                Path zipFile = Paths.get( generatedFile.getParent( ).toString( ), _generator.getFileName( ) );
+                FileUtil.zipFiles( zipFile, generatedFile );
+                physicalFile.setValue( Files.readAllBytes( zipFile ) );
+                Files.delete( zipFile );
```
Comment (rzara) [inline src/java/fr/paris/lutece/portal/service/tempfile/TemporaryFileGeneratorService.java]: the zip file might not be deleted if the preceding line throws

File: src/java/fr/paris/lutece/portal/service/html/HtmlCleanerService.java [hunk 2]
```diff
@@ -57,18 +57,35 @@ private HtmlCleanerService( )
      * @param strSource
      *            The input string to clean
      * @return The cleaned string
-     * @throws HtmlCleanerException
-     *             the HtmlCleanerException
      */
-    public static synchronized String clean( String strSource ) throws HtmlCleanerException
+    public static synchronized String clean( String strSource )
```
Comment (rzara) [inline src/java/fr/paris/lutece/portal/service/html/HtmlCleanerService.java]: This file does not seems to be part of this pull request

File: src/java/fr/paris/lutece/util/file/FileUtil.java [hunk 3]
```diff
@@ -146,4 +157,49 @@ public static void deleteFile( File file )
...
+    
+    /**
+     * Converts French diacritics characters into non diacritics. <br />
+     * Replace whitespaces by underscores <br />
+     * Keep only underscores and alphanum characters <br />
+     * Transform to lowercase
+     * @param string
+     * @return
+     */
+    public static final String normalizeFileName( String string )
```
Comment (rzara) [inline src/java/fr/paris/lutece/util/file/FileUtil.java]: This does not seem to be used in your PR. Maybe this should be a separate PR

File: src/java/fr/paris/lutece/portal/business/file/TemporaryFileHome.java [hunk 4]
```diff
@@ -140,4 +140,22 @@ public static TemporaryFile findByPrimaryKey( int nKey )
...
         return _dao.findByUser( user );
     }
+    
+    /**
+     * Delete the temporary files older than the given number of days
+     * @param days
+     */
+    public static void deleteFilesOlderThan( int days )
+    {
+        List<TemporaryFile> files = _dao.selectFilesOlderThan( days );
```
Comment (rzara) [inline src/java/fr/paris/lutece/portal/business/file/TemporaryFileHome.java]: Does this load the file data ? this seems wasteful.

File: src/java/fr/paris/lutece/portal/business/file/TemporaryFileHome.java [hunk 5]
```diff
@@ -140,4 +140,22 @@ public static TemporaryFile findByPrimaryKey( int nKey )
...
+        
+        for ( TemporaryFile file : files )
+        {
+            if ( file.getPhysicalFile( ) != null )
+            {
+                PhysicalFileHome.remove( file.getPhysicalFile( ).getIdPhysicalFile( ) );
+            }
+            _dao.delete( file.getIdFile( ) );
```
Comment (rzara) [inline src/java/fr/paris/lutece/portal/business/file/TemporaryFileHome.java]: These statements should be in a single transaction. If not, the physical file can be deleted when the tempory file isn't

File: src/java/fr/paris/lutece/portal/resources/tempfiles_messages_fr.properties [hunk 6]
```diff
@@ -14,3 +14,6 @@ manage_temporary_files.labelDate=Date de g\u00e9n\u00e9ration
 manage_temporary_files.state.done=G\u00e9n\u00e9ration termin\u00e9e
 manage_temporary_files.state.pending=G\u00e9n\u00e9ration en cours
 manage_temporary_files.state.error=Erreur de g\u00e9n\u00e9ration
+
+daemon.temporaryfilesDaemon.name=Daemon des fichiers temporaires
+daemon.temporaryfilesDaemon.description=Supprime les anciens fichiers tmporaires
```
Comment (rzara) [inline src/java/fr/paris/lutece/portal/resources/tempfiles_messages_fr.properties]: typo (manque un e)

File: src/java/fr/paris/lutece/portal/service/daemon/TemporaryFileDaemon.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * License 1.0
+ */
+package fr.paris.lutece.portal.service.daemon;
+
+import fr.paris.lutece.portal.business.file.TemporaryFileHome;
+import fr.paris.lutece.portal.service.util.AppPropertiesService;
+
+public class TemporaryFileDaemon extends Daemon
+{
+    private static final int DAYS_BEFORE_DELETE = Integer.parseInt( AppPropertiesService.getProperty( "tempfiles.days.defore.delete", "30" ) );
```
Comment (rzara) [inline src/java/fr/paris/lutece/portal/service/daemon/TemporaryFileDaemon.java]: You should query the property at the point of use, so that properties reload can have effect.

File: src/java/fr/paris/lutece/portal/business/file/ITemporaryFileDAO.java [hunk 8]
```diff
@@ -86,4 +86,10 @@
      * @return the list of File
      */
     List<TemporaryFile> findByUser( AdminUser user );
+    
+    /**
+     * Select files older than the given number of days
```
Comment (rzara) [inline src/java/fr/paris/lutece/portal/business/file/ITemporaryFileDAO.java]: Is days sufficiently granular ? Maybe you can split the parameter into a value and a `java.util.concurrent.TimeUnit`. See for instance https://github.com/lutece-platform/lutece-core/blob/dcb42f0ac5f9369104b5d72b6f175b81d0e94793/src/java/fr/paris/lutece/portal/service/mail/MailSenderDaemon.java#L86

File: src/java/fr/paris/lutece/portal/business/file/TemporaryFileHome.java (hunk 4 above)
Comment (glepine1) [reply src/java/fr/paris/lutece/portal/business/file/TemporaryFileHome.java]: No only the physical file id

**Comment by glepine1** [comment]
Will be implemented in [New Plugin](https://github.com/lutece-platform/lutece-tech-plugin-filegenerator)
